### PR TITLE
feat: iou, f1 score

### LINKED ISSUE #1: Investigate Segmentation Evaluation Metrics

### PR DESCRIPTION
**Pull request by rameziophobia**
not tested yet but check it out please

### CODE REVIEW

File: utils/evaluator.py [hunk 1]
```diff
@@ -7,19 +9,70 @@ class Evaluator:
...
+
+    def init_progress_bar(self):
+        # each hash represents 2 % of the progress
+        if self.is_progress_bar:
+            toolbar_width = 50
+            sys.stdout.write(f"[{'_' * toolbar_width}]")
+            sys.stdout.flush()
+
+    def img_to_tensor(self, img):
+        img_tensor = torch.from_numpy(np.array(img))  # not sure if this is correct/optimal
```
Comment (rameziophobia) [inline utils/evaluator.py]: should the data loader return NumPy arrays/ torch tensors instead of a PIL image?

File: utils/evaluator.py [hunk 2]
```diff
@@ -7,19 +9,70 @@ class Evaluator:
...
+        return total_iou / len(data_loader), total_f_score / len(data_loader)
+
+    def iou(self):
+        return self.current_intersection / \
+               ((self.current_union - self.current_intersection) + self.epsilon)
+
+    def f_score(self):
+        return 2 * self.current_intersection / (self.total_pixels + self.epsilon)
+
+    def print_progress(self, index, length):
```
Comment (rameziophobia) [inline utils/evaluator.py]: the progress bar stuff can be isolated in a separate file

File: utils/evaluator.py (hunk 1 above)
Comment (rameziophobia) [reply utils/evaluator.py]: if not then this function should be in the dataLoader class

File: utils/evaluator.py [hunk 3]
```diff
@@ -7,19 +9,70 @@ class Evaluator:
...
-        is_progress_bar=True,
-    ):
-        pass
+    def __init__(self,
+                 model,
+                 use_gpu=True,
+                 is_progress_bar=True,
+                 epsilon=1e-9):
+        self.model = model
+        self.use_gpu = use_gpu
```
Comment (MohamedAliRashad) [inline utils/evaluator.py]: No need to store the `use_gpu` argument as you already determined the device that you will be using.

File: utils/evaluator.py (hunk 2 above)
Comment (MohamedAliRashad) [reply utils/evaluator.py]: Try `tqdm`, will left all the hard work from you

File: utils/evaluator.py [hunk 4]
```diff
@@ -7,19 +9,70 @@ class Evaluator:
...
+        self.is_progress_bar = is_progress_bar
+        self.epsilon = epsilon
+        self.device = "cuda" if torch.cuda.is_available() and use_gpu else "cpu"
 
     def __call__(self, data_loader):
-        pass
+        data_iter = iter(data_loader)
```
Comment (MohamedAliRashad) [inline utils/evaluator.py]: check first if the dataloader is a list or a single object and try to retrieve its dataset name to print after evaluation.

File: utils/evaluator.py (hunk 2 above)
Comment (MohamedAliRashad) [inline utils/evaluator.py]: print should be the main function of `__call__` while the metric function be seprate entites.

File: utils/evaluator.py [hunk 5]
```diff
@@ -7,19 +9,70 @@ class Evaluator:
...
+        self.init_progress_bar()
+        for index, (img, ground_truth) in enumerate(data_iter):
+            img_tensor = self.img_to_tensor(img)
+            prediction = self.model.forward(img_tensor)
+            self.current_intersection = torch.sum(prediction * ground_truth)
+            self.total_pixels = torch.sum(ground_truth) + torch.sum(prediction)
+            self.current_union = self.total_pixels - self.current_intersection
+            total_iou += self.iou()
+            total_f_score += self.f_score()
+            self.print_progress(index, len(data_loader))
```
Comment (MohamedAliRashad) [inline utils/evaluator.py]: printing the loss for every iteration is not a wise choice, only print the last metric

File: utils/evaluator.py [hunk 6]
```diff
@@ -7,19 +9,70 @@ class Evaluator:
...
+        self.device = "cuda" if torch.cuda.is_available() and use_gpu else "cpu"
 
     def __call__(self, data_loader):
-        pass
+        data_iter = iter(data_loader)
+        total_iou = 0
+        total_f_score = 0
+        self.init_progress_bar()
+        for index, (img, ground_truth) in enumerate(data_iter):
+            img_tensor = self.img_to_tensor(img)
```
Comment (MohamedAliRashad) [inline utils/evaluator.py]: No need for this

File: utils/evaluator.py [hunk 7]
```diff
@@ -7,19 +9,70 @@ class Evaluator:
...
+            prediction = self.model.forward(img_tensor)
```
Comment (MohamedAliRashad) [inline utils/evaluator.py]: you can make it `model(img)` as forward is not required

File: utils/evaluator.py (hunk 4 above)
Comment (MohamedAliRashad) [reply utils/evaluator.py]: you can use this peace of code
```
if type(data_loader) is not list:
            data_loader = [data_loader]
```

**Review by MohamedAliRashad — CHANGES_REQUESTED** [review]

File: utils/evaluator.py (hunk 2 above)
Comment (rameziophobia) [reply utils/evaluator.py]: I tried not to use third-party libraries to avoid having unnecessary dependencies but using it will provide more stuff (ex: ETA) and it is easier. I'll try it

Comment (rameziophobia) [reply utils/evaluator.py]: I don't understand

File: utils/evaluator.py (hunk 4 above)
Comment (rameziophobia) [reply utils/evaluator.py]: why? it will work either way if it was a list or a single object

Comment (rameziophobia) [reply utils/evaluator.py]: > try to retrieve its dataset name to print after evaluation.  

I think that's not the responsibility of this class. this class computes evaluation metrics and its user can print the computation results and any metadata as he sees fit. (solid SRP)

Comment (MohamedAliRashad) [reply utils/evaluator.py]: True, but we want to encapsulate the evaluation process in one take ... where the user only passes the model and test_loaders and worry about nothing else.

The reason for the dataset name is for the tabular printing 
```
=================================
dataset_name | IoU | F1 Score | DICE | etc ....
=================================
```

File: utils/evaluator.py (hunk 2 above)
Comment (MohamedAliRashad) [reply utils/evaluator.py]: We wanted the class to have static functions for the metrics to be available if someone needed them outside the class while providing them to the `__call__` function for using and adding the tabular output

File: utils/evaluator.py [hunk 8]
```diff
@@ -30,16 +28,14 @@ def __call__(self, data_loader):
...
-            prediction = self.model.forward(img_tensor)
+        iterable = tqdm(enumerate(data_iter)) if self.is_progress_bar else enumerate(data_iter)
+        for index, (img, ground_truth) in iterable:
+            prediction = self.model(img)
             self.current_intersection = torch.sum(prediction * ground_truth)
             self.total_pixels = torch.sum(ground_truth) + torch.sum(prediction)
             self.current_union = self.total_pixels - self.current_intersection
             total_iou += self.iou()
             total_f_score += self.f_score()
-            self.print_progress(index, len(data_loader))
```
Comment (MohamedAliRashad) [inline utils/evaluator.py]: you can replace the print function with `texttable`, will do some heavy lifting too

**Review by MohamedAliRashad — CHANGES_REQUESTED** [review]

File: utils/evaluator.py (hunk 8 above)
Comment (rameziophobia) [reply utils/evaluator.py]: I removed that method. it was used to display the progress bar, sure

File: utils/evaluator.py (hunk 2 above)
Comment (rameziophobia) [reply utils/evaluator.py]: added static methods, made the other methods private. I'm using these private methods instead of the static ones because they use the same computations so it will be more efficient to calculate these only once.

File: utils/evaluator.py (hunk 4 above)
Comment (rameziophobia) [reply utils/evaluator.py]: now \_\_call\_\_ will add rows to a texttable and after doing several calls with different datasets the user may invoke print_metrics() to display the table

File: utils/evaluator.py [hunk 9]
```diff
@@ -23,6 +24,10 @@ def __init__(self,
         self.is_progress_bar = is_progress_bar
         self.epsilon = epsilon
```
Comment (MohamedAliRashad) [inline utils/evaluator.py]: I can't get my head why you made epsilon as an input parameter, Wouldn't it be better to have it as a constant small number and that's it.

File: utils/evaluator.py [hunk 10]
```diff
@@ -34,17 +39,40 @@ def __call__(self, data_loader):
...
-    def f_score(self):
+    def __f_score(self):
         return 2 * self.current_intersection / (self.total_pixels + self.epsilon)
 
+    def add_row_to_table(self):
+        self.table.add_rows([["", self.mean_iou, self.mean_f_score]])
+        print(self.table.draw())
+
+    def print_metrics(self):
+        print(self.table.draw())
```
Comment (MohamedAliRashad) [inline utils/evaluator.py]: There is no clear usage for these functions, they can be part of `__call__` and be printed after the evaluation process ends.

File: utils/evaluator.py [hunk 11]
```diff
@@ -34,17 +39,40 @@ def __call__(self, data_loader):
...
+        self.add_row_to_table()
 
-    def iou(self):
+    def __iou(self):
         return self.current_intersection / \
                ((self.current_union - self.current_intersection) + self.epsilon)
 
-    def f_score(self):
+    def __f_score(self):
         return 2 * self.current_intersection / (self.total_pixels + self.epsilon)
```
Comment (MohamedAliRashad) [inline utils/evaluator.py]: You already made the static functions that handle them, why is the rewriting ?

**Review by MohamedAliRashad — CHANGES_REQUESTED** [review]
Keep up the good work

File: utils/evaluator.py (hunk 11 above)
Comment (rameziophobia) [reply utils/evaluator.py]: to avoid computing the intersection and union twice thus taking twice the time to compute the metrics

**Review by MohamedAliRashad — APPROVED** [review]
